Epic refactoring because whathefuk

alexAubin opened the pull request:
So yeah I'm like at "only" 20hours of work on this and this is starting to look somewhat readable / evolvable ... I'm aiming to have a few separate pieces that can somewhat work independently from each other and that we can make evolve ... instead of a big pile of obscure global variables that magically change states in a completely incomprehensible, unnecessarily complex way. I'm just amazed that somehow the current code works so well...

yalh76 reviewed (approved):
it's like a yolo approval, I didn't read one line :( but trust you

alexAubin commented:
Sooooooo it's somewhat starting to work and ready for feedback ... Still need to test all use cases on the battlefield

There's so many things I changed and I'm a bit tired to write an exhaustive list but here's a try ... : 
- Ended up **switching to LXD instead of vanilla LXC** because I was fed up about all the "lol let's configure the network ourself with a bunch of iptable rules, this is so fun ! Also let's tweaks the rootfs manually all the time !". LXD is just way easier to work with (c.f. discussion below about having multiple package check in parallel for different test channels) ... Also this brings more consistency with ynh-dev and core ci, + possibly opens the possibly to install on non-Debian systems
- [Cleaned the output to be more human-friendly and focus on the actual test logic](https://user-images.githubusercontent.com/4533074/100564123-62585d80-32c0-11eb-8484-b958f77063e3.png)
- Tried to reorganize everything such that it's less of a big clumsy, repetitive bash soup. Tests are now ~somewhat independent of themselves and with more work could really be ran as standalone-entities I guess (in terms of not-relying-on-7548-magic-global-variables-to-work). Or at least things should be slightly more readable ...
- Rewrote the end summary / level analysis using python
- Parsing check_process is a lot more simpler now ... I guess could be turned into python but meh
- Dropped notifications stuff
- Dropped a bunch of weird options that were apparently not that useful
- Dropped a bunch of weird mechanism about "false positive" that imho was not doing anything useful...
- App script progression is displayed in real time
- The configpanel / actions tests are probably heavily broken at this point ... not sure I want to fix those considering the corresponding code is hell + things are still likely to move soon + only a handful of apps actually implement them ... Considering disablign them
- About check_process config file, turns out there no point in providing lines for args `domain`, `path`, `is_public` and `admin` ... (just a random thing about the simplifying-check_process-format-some-day™ .. as discussed in an issue on the bugtracker, most stuff could be guessed from the manifest)
- ...

### About tests

To give you an idea, the test summary for wordpress looks like (forget about the actual OK/fail results, this is just to discuss the list of test being ran) :  

```
Package linter:           OK 
Install (root):           OK 
Install (subdir):         OK 
Install (private):        OK 
Install (multi):         fail
Backup/restore:           OK 
Upgrade:                  OK 
Upgrade (00a1a6e7):      fail
Change url:              fail
Config/panel (actions):  fail
Config/panel (config_p): fail
Install (subdir):         OK 
```

In particular: 
- no more tests about "removal", they are included in the install tests..
- no more tests about "public install" because all install tests are public by default (apart from private install ofc)
- upgrade tests are independent from each other
- less madness about multiple test series .. in fact only a handful of apps have multiple test series. In my example, [the last "Install (subdir)" corresponds to an install from the alternative test serie](https://github.com/YunoHost-Apps/wordpress_ynh/blob/testing/check_process#L44) (this should be made explicit in the display...). Now tests from extra test series are only considered for a) root/subdir/nourl installs, b) upgrades from commit != current. For example, no test about backup/restore or private install is performed in extra test series because it does feel worth in terms of cost vs. gain ...

### About the configuration, LXD, CI perspectives

By default there's no `config` file anymore. Default values are defined are the top of `lib/common.sh`. Though you may still create a `config` file to override the default. In particular, one could imagine setting `ARCH=arm` and who knows ... if you rebuild the base image, it should pick up the ARM LXC and run everything with that arch (supposedly cross-architecture, meaning from an x64 machine). Absolutely not tested though...

Similarly, one could tweak the branch that's used for yunohost install and create a base image that's based on testing or unstable or next-debian...  Theoretically, one could run multiple instances of package_check as long as LXC_BASE and LXC_NAME are named appropriately to not conflict with other stuff...

### Possible other improvements

- [ ] Fix display of other-test-serie tests ... 
- [ ] Make sure Complete.log contains all the useful info for debug
- [ ] Try running tests on a non-web app
- [ ] Having a `--single-test <some_test_name>` option to run a single test instead of the entire test suite
- [ ] Having a `--self-upgrade` option to interface with CI_package_check (replacement for the auto_upgrade script) ?

kay0u commented on build_base_lxc.sh:
```diff
@@ -0,0 +1,53 @@
+    
+    INSTALL_SCRIPT="https://install.yunohost.org/$DIST"
+    $IN_LXC apt install curl -y
+    $IN_LXC /bin/bash -c "curl $INSTALL_SCRIPT | bash -s -- -a $YNH_BRANCH"
```
`$YNH_BRANCH` is not defined

alexAubin replied:
Indeed .. that was supposed to be https://github.com/YunoHost/package_check/blob/cleanup-3/lib/common.sh#L7 but forgot to rename stuff ... Ideally also should add a small tweak to automatically prefix `-d` if that var is not empty

kay0u replied:
we can use something like:

```
YNH_BRANCH=${1:-stable}
```

kay0u commented on lib/lxc.sh:
```diff
@@ -0,0 +1,166 @@
+        sudo lxc snapshot $LXC_NAME $snapname
+    fi
+
+    sudo lxc start --timeout 15 $LXC_NAME 2>/dev/null
```
```suggestion
    _LXC_START_AND_WAIT $LXC_NAME
```

kay0u commented:
I'd be nice to add an argument to be able to give a specific config file. and a

kay0u commented:
Backup/restore tests should be for both root and subpath (if the corresponding test was performed before)

kay0u commented on package_check.sh:
```diff
@@ -1,1518 +1,181 @@
+    if [ -n "$package_check_pid" ]; then
+        kill --signal 15 $package_check_pid
+    fi
+    clean_exit 0
```
We don't really want to exit the script here, just clean things up
```suggestion
    LXC_RESET

    [ -n "$TEST_CONTEXT" ] && rm -rf "$TEST_CONTEXT"
    rm -f "$lock_file"
```

alexAubin replied:
weeeell in fact this option is only meant to interface with CI_package_check which wants to clean things up if package_check times out ...

kay0u replied:
ho we're not supposed to use it with a package to test as agrument?

alexAubin replied:
eeeeh not really, at least that wasn't the intention xD

kay0u replied:
Ok! :-)

kay0u commented on README.md:
````diff
@@ -37,36 +34,130 @@ For a package on GitHub: `./package_check.sh https://github.com/YunoHost-Apps/AP
+The base images for tests are centralized on `devbaseimgs.yunohost.org` and we'll download them from there to speed things up:
+
+```bash
+# Stupid hack because ipv6 is interfering
+echo "$(dig +short A devbaseimgs.yunohost.org | tail -n 1) devbaseimgs.yunohost.org" >> /etc/hosts
````
```suggestion
```

Fixed now

kay0u commented on lib/lxc.sh:
```diff
@@ -0,0 +1,181 @@
+
+LXC_RESET () {
+    timeout 30 lxc stop --timeout 15 $LXC_NAME --force 2>/dev/null
+    swapoff $(cat /proc/swaps | grep "ynh-appci-buster-amd64-stable-test/rootfs/swap_") 2>/dev/null
```
```suggestion
    swapoff /var/snap/lxd/common/lxd/containers/$LXC_NAME/rootfs/swap_* 2>/dev/null
```
